_posts/2024-02-05-cde-customization.adoc

```diff
@@ -0,0 +1,434 @@
+
+When it's not possible to add a `.devfile.yaml` in the git repository, there is still the option to publish it (in a gist, pastebin or another git repository) and https://eclipse.dev/che/docs/stable/end-user-guide/starting-a-workspace-from-a-raw-devfile-url/[start the CDE providing the raw URL to it].
+
+TIP: To change the configuration of a running CDE, add a new `.devfile.yaml` (or edit the existing one), and restart the CDE using the command `Eclipse Che: Restart Workspace from Local Devfile` from Visual Studio Code command palette. The command is also quickly accessible by clicking on the bottom-left corner of Visual Studio Code.
+
+TIP: https://github.com/devfile/vscode-walkthrough-extension[The Devfile Visual Studio Code walkthrough extension] helps generating a Devfile using a GUI wizard.
+
+### Step by step guide to an optimal Devfile
+
+The https://devfile.io/docs/2.2.2/what-is-a-devfile[documentation website] includes a detailed Devfiles syntax reference. In this section I will share a practical step by step guide to write an optimal Devfile for the https://github.com/l0rd/rails-sample/[`rails-sample`].
+
+++++
+include::/assets/html/cde-customization/devfile-table.html[]
+++++
+{nbsp} +
+
+These steps can be tested iteratively from a running CDE using the "Eclipse Che: Restart Workspace from Local Devfile" in VS Code. And with the YAML extension you get automatic code completion when editing the Devfile.
+
+After the last step, the CDE will be fully functional. After the Devfile has been pushed to the git repository any new contributor will be able to run the application, do some changes, and debug them in a flash. 
+
+For example you can use https://workspaces.openshift.com/#https://github.com/l0rd/rails-sample/tree/devfile[this link] to test the rails-sample using the Devfile in step 7. After the workspace has started, use task `server-start` in VS Code (Terminal => Run Tasks => Devfile) to start the application and install the VS extension `ruby-lsp` to start making changes with full language support.
+
+## PART4: Secrets and other CDE configurations that cannot be specified in Devfile [[part4]]
+
+There are some properties of your CDE that you don't want to specify in a Devfile. Either because they contain sensible information (like a password or an SSH private key), because you want to customize your CDE without affecting the rest of the team or because that's an IDE configuration. In any case, the Devfile that is a shared file in the git repository, cannot be used. In this section we will go through a few techniques to add user specific configurations.
```

Comment:
```suggestion
There are some properties of your CDE that you don't want to specify in a Devfile. Either because they contain sensitive information (like a password or an SSH private key), because you want to customize your CDE without affecting the rest of the team or because it's an IDE configuration. In any case, the Devfile which is a shared file in the git repository, cannot be used. In this section we will go through a few techniques to add user specific configurations.
```